docs/codeexamples/writing_edsl.py

```diff
@@ -0,0 +1,441 @@
+        return O
+
+    def csum(I):
+        N = TensorDim()
+        i, k = TensorIndexes(2)
+        I.bind_dims(N)
+        O = TensorOutput(N)
+        O[i] += I[k]
+        O.add_constraint(i - k < N)
+        return O
+
+    def conv_1d(I, K):
+        N, X, KX, CI, CO = TensorDims(5)
+        n, x, k, ci, co = TensorIndexes(5)
+        I.bind_dims(N, X, CI)
+        K.bind_dims(KX, CI, CO)
+        O = TensorOutput(N, X - KX + 1, CO)
+        O[n, x, co] += I[n, x + k, ci] * K[k, ci, co]
+        return O
+
+    def conv_2d_dilated(I, K):
+        N, X, Y, KX, KY, CI, CO = TensorDims(7)
+        n, x, y, kx, ky, ci, co = TensorIndexes(7)
+        I.bind_dims(N, X, Y, CI)
+        K.bind_dims(KX, KY, CI, CO)
+        O = TensorOutput(N, X - 2 * (KX - 1), Y - 3 * (KY - 1), CO)
+        O[n, x, y, co] += I[n, x + 2 * kx, y + 3 * ky, ci] * K[kx, ky, ci, co]
+        return O
+
+    def complex_conv_2d(I, K, s0, s1, d0, d1):
+        # "same-lower" autopadding will be applied
+        N, G, GCI, GCO = TensorDims(4)
+        X0, X1 = TensorDims(2)
+        K0, K1 = TensorDims(2)
+        n, g, gci, gco = TensorIndexes(4)
+        x0, x1 = TensorIndexes(2)
+        k0, k1 = TensorIndexes(2)
+        I.bind_dims(N, X0, X1, G, GCI)
+        K.bind_dims(K0, K1, G, GCI, GCO)
+
+        # Compute output spatial dimensions
+        Y0, Y1 = TensorDims(2)
+        Y0 = (X0 + s0 - 1) // s0
+        Y1 = (X1 + s1 - 1) // s1
+
+        #Compute the effective kernel size after dilation
+        EK0, EK1 = TensorDims(2)
+        EK0 = d0 * (K0 - 1) + 1
+        EK1 = d1 * (K1 - 1) + 1
+
+        #Compute the padding offset
+        P0, P1 = TensorDims(2)
+        P0 = ((Y0 - 1) * s0 + EK0 - X0) // 2
+        P1 = ((Y1 - 1) * s1 + EK1 - X1) // 2
+
+        # Specify the output size
+        O = TensorOutput(N, Y0, Y1, G, GCO)
+
+        # Compute the convolution
+        O[n, x0, x1, g, gco] += I[n, s0 * x1 + d0 * k0 - P0, s1 * x1 + d1 * k1 -
+                                  P1, g, gci] * K[k0, k1, g, gci, gco]
+        return O
+
+
+class TestEdslDocs(unittest.TestCase):
+    maxDiff = None
+
+    def compare_results(self, program, expected):
+        self.assertMultiLineEqual(str(program).strip(), expected.strip())
+
+    def test_sum_over_axis(self):
+        I = Tensor(LogicalShape(plaidml.DType.FLOAT32, [1, 78]))
+        O = TestEdslHelper.sum_over_axis(I)
+        program = Program('sum_over_axis', [O])
+        expected = '''
+#map0 = affine_map<(d0, d1) -> (d0)>
+#map1 = affine_map<(d0, d1) -> (d1, d0)>
+
+
+!f32 = type tensor<!eltwise.f32>
+module {
+  func @sum_over_axis(%arg0: tensor<1x784x!eltwise.f32>) -> tensor<784x!eltwise.f32> {
+    %cst = "eltwise.sconst"() {value = 0.000000e+00 : f64} : () -> !f32
+    %0 = tile.cion add, none, %cst, %arg0 {sink = #map0, srcs = [#map1]} : !f32, tensor<1x784x!eltwise.f32> -> tensor<784x!eltwise.f32>
+    return %0 : tensor<784x!eltwise.f32>
+  }
+}
+'''
+        self.compare_results(program, expected)
```

Comment:
Drop the `self.compare_results` for all for test tests. We only need to verify that it compiles at all, not that it generates this exact sequence of MLIR code.